File: sieve_benchmark/prime.py
import sys
import time

def sieve(n: int) -> int:
    is_prime = [True] * (n + 1)
    for i in range(2, int(n ** 0.5) + 1):
        if is_prime[i]:
            for j in range(i * i, n + 1, i):
                is_prime[j] = False
    return sum(1 for i in range(2, n + 1) if is_prime[i])

def main():
    if len(sys.argv) < 2:
        return
    n = int(sys.argv[1])
    start = time.time()
    count = sieve(n)
    elapsed = time.time() - start
    print(f"primes={count} time={elapsed:.6f}")

if __name__ == "__main__":
    main()

File: sieve_benchmark/prime.js
const n = parseInt(process.argv[2]);
if (!n) process.exit(1);
const isPrime = Array(n + 1).fill(true);
const start = process.hrtime.bigint();
for (let p = 2; p * p <= n; p++) {
  if (isPrime[p]) {
    for (let j = p * p; j <= n; j += p) {
      isPrime[j] = false;
    }
  }
}
let count = 0;
for (let i = 2; i <= n; i++) {
  if (isPrime[i]) count++;
}
const end = process.hrtime.bigint();
const elapsed = Number(end - start) / 1e9;
console.log(`primes=${count} time=${elapsed.toFixed(6)}`);
